### ISSUE TITLE
Generalize strip_truncations to strip_reflections?

### ISSUE THREAD

**Issue by jdchristensen** [HoTT/HoTT#1481]
Here's a handy result.  At first, you might think it needs `Funext`, since that is needed to show that `Y -> Z` is O-local when `Z` is O-local.  But by juggling the context, you can avoid `Funext`.

```coq
Lemma O_concat {O : ReflectiveSubuniverse} {A : Type} {a0 a1 a2 : A}
  : O (a0 = a1) -> O (a1 = a2) -> O (a0 = a2).
Proof.
  intro Op.
  rapply O_rec; intro q.
  revert Op; rapply O_rec; intro p.
  exact (to O _ (p @ q)).
Defined.
```

This can be abstracted.  E.g. here's something that just handles the first argument.

```coq
Definition O_rec_to_arrow (O : ReflectiveSubuniverse) {X Y Z : Type} `{In O Z}
           (f : X -> (Y -> Z))
  : (O X) -> (Y -> Z).
Proof.
  intros ox y.
  revert ox.
  apply O_rec.
  intro x.
  exact (f x y).
Defined.

Definition O_rec_to_arrow_beta (O : ReflectiveSubuniverse) {X Y Z : Type} `{In O Z}
           (f : X -> (Y -> Z)) (x : X)
  : (O_rec_to_arrow O f) ((to O X) x) == (f x).
Proof.
  intro y.
  rapply O_rec_beta.
Defined.
```

With this, one can write:

```coq
Lemma O_concat' {O : ReflectiveSubuniverse} {A : Type} {a0 a1 a2 : A}
  : O (a0 = a1) -> O (a1 = a2) -> O (a0 = a2).
Proof.
  rapply O_rec_to_arrow; intro p.
  rapply O_rec; intro q.
  exact (to O _ (p @ q)).
Defined.
```

Even better, maybe we should generalize `strip_truncations` to `strip_reflections`:

```coq
(** A tactic that generalizes strip_truncations. *)
Ltac strip_reflections :=
  (** Search for hypotheses of type [O X] for some [O] such that the goal is [O]-local. *)
  progress repeat match goal with
                    | [ T : _ |- _ ]
                      => revert_opaque T;
                        (** The dependent case requires that [O] be a modality. *)
                        refine (@O_rec _ _ _ _ _ _ _) || refine (@O_ind _ _ _ _ _ _ _);
                        (** Ensure that we didn't generate more than one subgoal, i.e. that the goal was appropriately local. *)
                        [];
                        intro T
                  end.

Lemma O_concat'' {O : ReflectiveSubuniverse} {A : Type} {a0 a1 a2 : A}
  : O (a0 = a1) -> O (a1 = a2) -> O (a0 = a2).
Proof.
  intros p q.
  strip_reflections.
  exact (to O _ (p @ q)).
Defined.
```

Questions:

- Is `strip_reflections` likely to be useful?  If so, where should it be put?  Should it be separated into a version for Modalities and a version for ReflectiveSubuniverses?
- Would adding `O_concat` and other operations on paths be useful?
- I thought `strip_reflections` might be able to replace `strip_truncations`.  It does work in some cases, but it gives "failed to progress" or "Universe inconsistency" in other cases, e.g. in Projective.v.  Not sure what is going on there.

PS:  Does the above show that `Y -> Z` *is* O-local in the wild category of types (where equality of morphisms is homotopy), when `Z` is O-local, without using `Funext`?

**Comment by mikeshulman** [comment]
I don't know if these things would be useful, but it couldn't hurt to add them in case they turn out to be.

I'm not sure about the last question, but it does sound like it.

**Comment by Alizter** [comment]
@jdchristensen I've made a `strip_reflections` and `strip_modalities` tactic which call O_rec and O_ind respectively. The latter doesn't change the universe where the former does leading to problems in Projective.v.

**Comment by jdchristensen** [comment]
Fixed by #1546.
